library/ssl_tls.c

```diff
@@ -8124,43 +8223,41 @@ static size_t ssl_tls12_session_save( const mbedtls_ssl_session *session,
      * Session ticket if any, plus associated data
      */
 #if defined(MBEDTLS_SSL_SESSION_TICKETS) && defined(MBEDTLS_SSL_CLI_C)
-    used += 3 + session->ticket_len + 4; /* len + ticket + lifetime */
+    /* ticket_len                       3 bytes
+     * ticket                  ticket_len bytes
+     * lifetime                         4 bytes
+     */
+    MBEDTLS_SSL_CHK_BUF_PTR( p, end, 3 );
+    MBEDTLS_PUT_UINT24_BE( session->ticket_len, p, 0 );
+    p += 3;
 
-    if( used <= buf_len )
+    MBEDTLS_SSL_CHK_BUF_PTR( p + 3, end, session->ticket_len );
+    if( session->ticket != NULL )
     {
-        *p++ = MBEDTLS_BYTE_2( session->ticket_len );
-        *p++ = MBEDTLS_BYTE_1( session->ticket_len );
-        *p++ = MBEDTLS_BYTE_0( session->ticket_len );
+        memcpy( p, session->ticket, session->ticket_len );
+    }
+    p += session->ticket_len;
 
-        if( session->ticket != NULL )
-        {
-            memcpy( p, session->ticket, session->ticket_len );
-            p += session->ticket_len;
-        }
+    MBEDTLS_SSL_CHK_BUF_PTR( p, end, 4 );
+    MBEDTLS_PUT_UINT32_BE( session->ticket_lifetime, p, 0 );
+    p += 4;
 
-        MBEDTLS_PUT_UINT32_BE( session->ticket_lifetime, p, 0 );
-        p += 4;
-    }
 #endif /* MBEDTLS_SSL_SESSION_TICKETS && MBEDTLS_SSL_CLI_C */
 
     /*
      * Misc extension-related info
      */
 #if defined(MBEDTLS_SSL_MAX_FRAGMENT_LENGTH)
-    used += 1;
-
-    if( used <= buf_len )
-        *p++ = session->mfl_code;
+    MBEDTLS_SSL_CHK_BUF_PTR( p, end, 1 );
+    *p++ = session->mfl_code;
 #endif
 
 #if defined(MBEDTLS_SSL_ENCRYPT_THEN_MAC)
-    used += 1;
-
-    if( used <= buf_len )
-        *p++ = MBEDTLS_BYTE_0( session->encrypt_then_mac );
+    MBEDTLS_SSL_CHK_BUF_PTR( p, end, 1 );
+    *p++ = MBEDTLS_BYTE_0( session->encrypt_then_mac );
 #endif
 
-    return( used );
+    return( 0 );
```

Comment:
Curius as to why we setting the `MBEDTLS_CHECK_RETURN_CRITICAL` attribute and then returning always 0?

Reply:
`MBEDTLS_SSL_CHK_BUF_PTR` will return buffer too small error.